File: generated_code/optimized/runtime/java/7.java
import java.util.*;

class Solution {
    /**
    Filter an input list of strings only for ones that contain given substring
    >>> filterBySubstring(List.of(), "a")
    []
    >>> filterBySubstring(Arrays.asList("abc", "bacd", "cde", "array"), "a")
    ["abc", "bacd", "array"]
     */
    public List<String> filterBySubstring(List<String> strings, String substring) {
        int n = strings.size();
        if (n == 0) return Collections.emptyList();
        List<String> result = new ArrayList<>(n);
        if (strings instanceof RandomAccess) {
            for (int i = 0; i < n; i++) {
                String s = strings.get(i);
                if (s.contains(substring)) result.add(s);
            }
        } else {
            for (String s : strings) {
                if (s.contains(substring)) result.add(s);
            }
        }
        return result;
    }
}

File: generated_code/optimized/memory/python/7.py
from typing import List


def filter_by_substring(strings: List[str], substring: str) -> List[str]:
    """ Filter an input list of strings only for ones that contain given substring
    >>> filter_by_substring([], 'a')
    []
    >>> filter_by_substring(['abc', 'bacd', 'cde', 'array'], 'a')
    ['abc', 'bacd', 'array']
    """
    write = 0
    for read in range(len(strings)):
        s = strings[read]
        if substring in s:
            strings[write] = s
            write += 1
    del strings[write:]
    return strings
